<a href="https://colab.research.google.com/github/izzul112/ML_BMLT-Proyek_Akhir-Sistem_Rekomendasi-Anime/blob/main/Proyek_Akhir_Sistem_Rekomendasi.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Laporan Proyek Machine Learning - Habib Azizul Haq**

# Project Overview

Anime sebuah film animasi dari Jepang yang sangat populer dan mungkin banyak dari sebagian kalian yang merasa nostalgia ketika waktu kecil sering menontonnya di televisi. Sekarang untuk menontonnya banyak di antara kita yang sering mengunjungi situs-situs *streaming* seperti Bsite, IQiy, atau bahkan di kanal Youtube seperti Ani-One, Muse Indonesia dan Muse Asia.

Kepopuleran Anime saat memang lagi tinggi-tingginya, hal ini disebabkan banyak judul-judul besar Anime yang sudah punya basis penggemar yang besar, di musim panas 2022 ini banyak yang mendapat lanjutan musim keduanya bahkan Anime dengan judul Overlord mendapat kelanjutan musim keempat, dimana kelanjutan series ini sangat di tunggu-tunggu, perilisan musim keempat ini juga di iringi dengan rilisnya volume terbaru dari Light Novel Overlord yaitu volume 15 dan 16, tentu ini makin menambah *hype* dari judul Anime ini.


![image](https://user-images.githubusercontent.com/43197282/182755994-e4e082a1-d997-4b59-bffd-f675c4973c1b.png)
![image](https://user-images.githubusercontent.com/43197282/182756138-d9a75610-75c8-4c5a-a645-f3c4694ab197.png)
![image](https://user-images.githubusercontent.com/43197282/182756296-1afc570b-09c7-456c-a2ce-12fa73adf741.png)


Melihat tingginya kepopuleran yang terjadi sekarang juga membawa berkah tersendiri bagi industri *streaming* untuk medulang cuan yang banyak. Dengan adanya situs-situs streaming seperti disebutkan diatas membantu kita sebagai orang yang sedang mencari Anime untuk menonton secara **Legal**. Lantas ketika Anda mengunjungi katakanlah kanal youtube Muse Indonesia dan meng-klik salah-satu judul Anime disana, Anda akan melihat tampilan video lain yang serupa ketika Anda menggeser kebawah. Pernahkan Anda berfikir bagaimana video dibawah tersebut bisa muncul? Apa mereka muncul secara acak? *Tentu tidak demikian ferguso*.

Video dibawah muncul melalui serangkaian algoritma yang sering di sebut dengan sistem rekomendasi, ia merekomendasikan sebuah video yang mungkin menarik bagi Anda, supaya Anda meng-klik video tersebut. Lantas sekarang bayangkan Anda orang di bagian marketing dari salah-satu situs *streming* diatas, tentu Anda ingin meningkatkan pengalaman pengguna ketika menggunakan situs Anda, supaya pengguna merasa nyaman dan ingin berlangganan tetap di situs *streaming* Anda.

Sekarang Anda mengajukan sebuah proposal kepada atasan Anda untuk membuat sistem rekomendasi dan menerapkannya di situs *streaming*, Anda mengatakan alasan penting mengapa perlu sistem rekomendasi di situs *streaming* saat ini:

- Untuk meningkatkan kepusaan pengguna, mengingat setiap pengguna belum tentu tau banyak judul Anime.
- Untuk memahami yang lebih baik tentang preferensi pengguna.

Dengan harapan bahwa hal itu bisa meningkatkan keuntungan bagi perusahaan dan tentu saja jika ini sukses di implementasikan dan menarik lebih banyak pelanggan maka tentu gaji Anda akan di naikkana oleh perusahhana Anda, karena kontribusi Anda.

- [Referensi]()

# Business Understanding

**Problem Statements**

Menjelaskan pernyataan masalah latar belakang:

*  Anda belum memiliki sistem rekomendasi, yang memberikan rekomendasi-rekomendasi Anime yang mungkin di sukai oleh pengguna di situs *streaming* Anda.
*  Anda belum mengetahui preferensi dari pengguna situs *streaming* Anda.

**Goals**

Menjelaskan tujuan dari pernyataan masalah:

*   Anda ingin meningkatkan pengalaman pengguna dengan memberikan rekomendasi-rekomendasi Anime yang mungkin di sukai oleh pengguna di situs *streaming* Anda.
*   Anda ingin mengetahui preferensi dari pengguna situs *streaming* Anda.

**Solution statements**

*   Kita akan menggunkana pendekatan Sistem Rekomendasi: Content-based Filtering, untuk memberikan rekomendasi Anime.
*   Kita akan menggunakan model dengan metrik `cosine_similarity` untuk menentukan kemiripan dari suatu judul Anime dalam memberikan suatu hasil rekomendasi.
*   Kita akan menggunkana metrik `Precision` untuk menilai seberapa akurat hasil rekomendasi dari sistem rekomendasi yang kita buat.


## Data Understanding

Kita akan menggunkan dataset yang berisi 23 variabel yang sering muncul ketika Anda melihat deskripsi suatu Anime di situs MyAnimeList[1]. File dataset yang akan kita gunakan dalam proyek ini adalah anime_list.csv yang berisi 23 vaiabel dan 18.000 judul Anime.

Sumber dataset yang akan kita gunakan berasal dari [Kaggle](https://www.kaggle.com/datasets/snehaanbhawal/anime-list-for-recommendation-system-june-2021). Yang di publis oleh [SNEHAAN BHAWAL](https://www.kaggle.com/snehaanbhawal).

**Variabel-variabel pada *Anime List for Recommendation System* dataset adalah sebagai berikut:**

- mal_id : id unik dari situs myanimelist.net yang mengidentifikasi Anime.
- title : Judul dari Anime
- synopsis : Sinopsis, ringkasan singkat yang memberikan gambaran tentang suatu karya.
- background : Latar belakang.
- aired : Rentang waktu anime ditayangkan. contoh format: "30 Sep 2004 sampai 29 Sep 2005".
- airing : Apakah saat ini sedang ditayangkan atau tidak. 0 jika anime telah berakhir 1 jika sedang berlangsung
- duration : Durasi per-episode.
- episodes : Jumlah episode dari suatu judul Anime.
- type : Jenis Anime seperti: ['TV', 'Movie', 'OVA', 'Special', 'ONA', 'Music', 'Unknown']
- favorites : Jumlah orang yang memfavoritkan anime ini.
- members : Jumlah anggota.
- rank : Peringkat per Juni 2021.
- popularity : Peringkat Popularitas per Juni 2021.
- score : Skor rata-rata yang diberikan oleh pengguna.
- scored_by : Jumlah orang yang merating Anime.
- rating : Peringkat diberikan untuk anime.['PG - Anak', 'PG-13 - Remaja 13 tahun ke atas', 'G - Semua Usia', 'R+ - Ketelanjangan Ringan', 'R - 17+','Rx - Hentai' ]
- premiered : Tahun dan musim pemutaran perdana.
- genres : Genre Anime.
- related : Anime terkait.
- status : Status Anime apakah sedang tayang saat ini atau tidak.
- licensors : Nama Pemberi Lisensi.
- producers : Nama Produser.
- studios : Nama studio yang menggerjakan.

Kita akan melakukan sedikit analisis untuk mengethui beberapa hal tentang variabel diatas diantaranya :

1. Kita akan melihat ada berapa jumlah baris dari data dalam tampilan tabel.
2. Kita akan melihat tipe data dari 21 kolom di dataset.
3. Kita akan mengecek deskripsi statistik data.
4. Kita akan melihat apakah ada data yang kosong / tidak ada isinya.

Sebelum kita mulai mengerjakan kita import semua library yang di butuhkan.



In [1]:
# Import library
import pandas as pd
import numpy as np 
import tensorflow as tf
from tensorflow import keras
from keras import layers
from pathlib import Path
import matplotlib.pyplot as plt

Kita sambungkan dulu Google Colab kita dengan Google Drive, karena dataset kita berada di Google Drive.

In [2]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


Setelah itu kita buat variabel untuk menampung dataset dengan nama data_anime untuk menampung data anime_list.csv yang akan kita panggil dari Google Drive yang sudah kita sambungkan dengan Google Colab.



In [3]:
data_anime = pd.read_csv('/content/gdrive/My Drive/dataset/anime_list.csv')

# **Pembahasan**
**1. Kita akan melihat ada berapa jumlah baris dari data dalam tampilan tabel.**

Kita akan menggunakan menggunakan kode `data_anime`, `data_anime` adalah nama variabel yang kita ganakan saat meload dataset anime_list.csv.

In [4]:
data_anime

mal_id                                              title  \
0           1                                       Cowboy Bebop   
1           5                    Cowboy Bebop: Tengoku no Tobira   
2           6                                             Trigun   
3           7                                 Witch Hunter Robin   
4           8                                     Bouken Ou Beet   
...       ...                                                ...   
18157   48952              Twisted Wonderland 1st Anniversary PV   
18158   48953  Natsume Yuujinchou: Hitoyoshi Kuma de no Yasas...   
18159   48954                                     Hairpin Double   
18160   48956                        Wu Shang Shen Di 2nd Season   
18161   48960                                             Hunter   

                                                synopsis background  \
0      In the year 2071, humanity has colonized sever...          -   
1      Another day, another bounty—such is the life o...          -   
2      Vash the Stampede is the man with a $$60,000,0...          -   
3      Witches are individuals with special powers li...          -   
4      It is the dark century and the people are suff...          -   
...                                                  ...        ...   
18157  Disney's Twisted Wonderland mobile game had it...          -   
18158  A collaboration with the Kumamoto prefecture i...          -   
18159                                                  -          -   
18160                                                  -          -   
18161  Hunter is looking for Gossip every night. A sh...          -   

                              aired  airing       duration  episodes     type  \
0       Apr 3, 1998 to Apr 24, 1999       0  24 min per ep        26       TV   
1                       Sep 1, 2001       0    1 hr 55 min         1    Movie   
2       Apr 1, 1998 to Sep 30, 1998       0  24 min per ep        26       TV   
3       Jul 2, 2002 to Dec 24, 2002       0  25 min per ep        26       TV   
4      Sep 30, 2004 to Sep 29, 2005       0  23 min per ep        52       TV   
...                             ...     ...            ...       ...      ...   
18157                  Mar 17, 2021       0          2 min         1  Special   
18158                  Apr 21, 2021       0          2 min         1      ONA   
18159                     2021 to ?       0        Unknown        -1       TV   
18160             May 21, 2021 to ?       1   7 min per ep        64      ONA   
18161                  Sep 14, 2017       0          8 min         1    Movie   

       favorites  ...  score  scored_by                          rating  \
0          59968  ...   8.77     661519  R - 17+ (violence & profanity)   
1           1063  ...   8.39     168515  R - 17+ (violence & profanity)   
2          11882  ...   8.23     288760       PG-13 - Teens 13 or older   
3            512  ...   7.27      37135       PG-13 - Teens 13 or older   
4             10  ...   6.97       5463                   PG - Children   
...          ...  ...    ...        ...                             ...   
18157          0  ...  -1.00         -1       PG-13 - Teens 13 or older   
18158          0  ...  -1.00         -1                    G - All Ages   
18159          0  ...  -1.00         -1                            None   
18160          2  ...  -1.00         -1       PG-13 - Teens 13 or older   
18161          0  ...  -1.00         -1                R+ - Mild Nudity   

         premiered                                             genres  \
0      Spring 1998    Action, Adventure, Comedy, Drama, Sci-Fi, Space   
1                -              Action, Drama, Mystery, Sci-Fi, Space   
2      Spring 1998  Action, Sci-Fi, Adventure, Comedy, Drama, Shounen   
3      Summer 2002  Action, Mystery, Police, Supernatural, Drama, ...   
4        Fall 2004          Adventure, Fantasy, Shounen, Supernatural   
...            ...       

Dari kode `data_anime` kita mendapat informasi:

*   Ada 18.162 baris dataset anime_list.csv kita
*   Dan terdapat 23 kolom



**2. Kita akan melihat tipe data dari kedua dataset kita.**

Kita akan menggunakan `data_anime.info()`, `.info()` akan menampilkan informasi tipe data 7 kolom dataset anime.csv yang kita gunakan.

In [5]:
data_anime.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18162 entries, 0 to 18161
Data columns (total 23 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   mal_id      18162 non-null  int64  
 1   title       18162 non-null  object 
 2   synopsis    18162 non-null  object 
 3   background  18162 non-null  object 
 4   aired       18162 non-null  object 
 5   airing      18162 non-null  int64  
 6   duration    18162 non-null  object 
 7   episodes    18162 non-null  int64  
 8   type        18162 non-null  object 
 9   favorites   18162 non-null  int64  
 10  members     18162 non-null  int64  
 11  rank        18162 non-null  float64
 12  popularity  18162 non-null  int64  
 13  score       18162 non-null  float64
 14  scored_by   18162 non-null  int64  
 15  rating      18162 non-null  object 
 16  premiered   18162 non-null  object 
 17  genres      18095 non-null  object 
 18  related     18162 non-null  object 
 19  status      18162 non-nul

Dari *output* `data_anime.info()`, terlihat bahwa:

*   Terdapat 2 kolom dengan tipe data float64, yaitu: 
*   Terdapat 7 kolom dengan tipe data int64, yaitu: 
*   Terdapat 14 kolom dengan tipe data object, yaitu: 



**3. Kita akan mengecek deskripsi statistik data.**

Kita akan mengeceknya dengan fungsi `describe()`.

In [6]:
data_anime.describe()

mal_id        airing      episodes      favorites       members  \
count  18162.000000  18162.000000  18162.000000   18162.000000  1.816200e+04   
mean   22321.114470      0.017564     11.945436     424.283889  3.698543e+04   
std    15330.497714      0.131364     48.074396    3858.020730  1.372226e+05   
min        1.000000      0.000000     -1.000000       0.000000  1.100000e+01   
25%     6297.500000      0.000000      1.000000       0.000000  3.330000e+02   
50%    24086.000000      0.000000      2.000000       3.000000  2.012500e+03   
75%    36431.750000      0.000000     12.000000      27.000000  1.338775e+04   
max    48960.000000      1.000000   3057.000000  175658.000000  2.884844e+06   

               rank    popularity         score     scored_by  
count  18162.000000  18162.000000  18162.000000  1.816200e+04  
mean    7418.114029   9091.728554      4.301493  1.833681e+04  
std     5125.230109   5252.673268      3.483474  7.974844e+04  
min       -1.000000      1.000000     -1.000000 -1.000000e+00  
25%     2782.500000   4544.250000     -1.000000 -1.000000e+00  
50%     7328.000000   9087.000000      6.020000  5.850000e+02  
75%    11868.750000  13637.750000      6.830000  4.714000e+03  
max    16464.000000  18247.000000      9.170000  1.983096e+06

Fungsi `describe()` memberikan informasi statistik pada masing-masing kolom, antara lain:

*   Count  adalah jumlah sampel pada data.
*   Mean adalah nilai rata-rata.
*   Std adalah standar deviasi.
*   Min yaitu nilai minimum setiap kolom.
*   25% adalah kuartil pertama. Kuartil adalah nilai yang menandai batas interval dalam empat bagian sebaran yang sama.
*   50% adalah kuartil kedua, atau biasa juga disebut median (nilai tengah).
*   75% adalah kuartil ketiga.
*   Max adalah nilai maksimum.

Perhatikan pada kolom `score` bisa kita lihat bahwa baris Max atau nilai maksimum yang diberikan oleh rata-rata user adalah 9.17 dan baris Min yaitu nilai terendahnya adalah -1. Niali -1 terjadi ketika user melihat Anime tersebut tapi tidak memberikan rating.

**4. Kita akan melihat apakah ada data yang kosong / tidak ada isinya.**

Kita akan mengeceknya dengan fungsi `.isnull().sum()`. Fungsi `.isnull()` akan mengecek apakah ada data kosng pada setiap baris pada semua kolom di dataset kita, kemudian kita gunakan juga fungsi `.sum()` untuk menjumlahnya sehingga hasilnya akan seperti berikut.

In [7]:
data_anime.isnull().sum()

mal_id         0
title          0
synopsis       0
background     0
aired          0
airing         0
duration       0
episodes       0
type           0
favorites      0
members        0
rank           0
popularity     0
score          0
scored_by      0
rating         0
premiered      0
genres        67
related        0
status         0
licensors      0
producers      0
studios        0
dtype: int64

Dari hasil di atas kita tau bahwa di kolom genres memiliki 67 data kosong.

# **Data Preparation**

Setelah kita melalui proses Data Understanding, kita menjadi lebih paham tentang seperti apa isi dataset kita.

Oleh karena itu di proses ini kita akan melakukan 2 proses yaitu:

1.  Kita akan mengubah nilai score yang -1.
2.  Kita akan menghapus data yang kosong.

**1.  Kita akan mengubah nilai score yang -1.**

Penghapusan ini penting mengingat nilai -1 terjadi karena user menonton anime tersebut tapi tidak memberikan score, maka dari itu supaya nilai -1 ini tidak mempengaruhi sitem rekomendasi kita, kita akan menggantinya dengan nilai NaN / kosong supaya bisa kita hapus baris data yang mengandung nilai -1.

Sebelum itu kita lihat dulu nilai ada berapa jumlah nilai -1 nya.

Kita akan melihatnya dengan kode `data_anime.loc[(data_anime['score']==-1)]`.
- `.loc[]` adalah akses grub baris dan kolom meurut label, dimana kita ingin mengakses kolom score dari data_anime.
- `==-1` adalah parameter untuk hanya menampilkan kolom score dengan nilai yang sama dengan -1.

In [8]:
data_anime.loc[(data_anime['score']==-1)]

mal_id                                              title  \
1574     1739                         Shibawanko no Wa no Kokoro   
1695     1863                            Silk Road Shounen Yuuto   
1986     2173  Kouchuu Ouja Mushiking Super Battle Movie: Yam...   
2161     2366                                Touma Kishinden Oni   
2418     2636          Dorami-chan: Wow, The Kid Gang of Bandits   
...       ...                                                ...   
18157   48952              Twisted Wonderland 1st Anniversary PV   
18158   48953  Natsume Yuujinchou: Hitoyoshi Kuma de no Yasas...   
18159   48954                                     Hairpin Double   
18160   48956                        Wu Shang Shen Di 2nd Season   
18161   48960                                             Hunter   

                                                synopsis background  \
1574   Based on a japanese children`s book by Yoshie ...          -   
1695   When a boy Yuto visits Qinghai in China, he is...          -   
1986   A shadow of Darkness approaches a peaceful For...          -   
2161   Shuramaru is hated and feared by the villagers...          -   
2418   No synopsis has been added for this series yet...          -   
...                                                  ...        ...   
18157  Disney's Twisted Wonderland mobile game had it...          -   
18158  A collaboration with the Kumamoto prefecture i...          -   
18159                                                  -          -   
18160                                                  -          -   
18161  Hunter is looking for Gossip every night. A sh...          -   

                              aired  airing       duration  episodes     type  \
1574    Apr 5, 2006 to Mar 14, 2007       0   2 min per ep        80       TV   
1695   Sep 16, 2006 to Mar 24, 2007       0  20 min per ep        26       TV   
1986                   Mar 21, 2007       0         50 min         1    Movie   
2161    Oct 5, 1995 to Mar 21, 1996       0   8 min per ep        25       TV   
2418                    Mar 9, 1991       0         40 min         1    Movie   
...                             ...     ...            ...       ...      ...   
18157                  Mar 17, 2021       0          2 min         1  Special   
18158                  Apr 21, 2021       0          2 min         1      ONA   
18159                     2021 to ?       0        Unknown        -1       TV   
18160             May 21, 2021 to ?       1   7 min per ep        64      ONA   
18161                  Sep 14, 2017       0          8 min         1    Movie   

       favorites  ...  score  scored_by                     rating  \
1574           0  ...   -1.0         -1               G - All Ages   
1695           0  ...   -1.0         -1                       None   
1986           1  ...   -1.0         -1               G - All Ages   
2161           0  ...   -1.0         -1  PG-13 - Teens 13 or older   
2418           0  ...   -1.0         -1               G - All Ages   
...          ...  ...    ...        ...                        ...   
18157          0  ...   -1.0         -1  PG-13 - Teens 13 or older   
18158          0  ...   -1.0         -1               G - All Ages   
18159          0  ...   -1.0         -1                       None   
18160          2  ...   -1.0         -1  PG-13 - Teens 13 or older   
18161          0  ...   -1.0         -1           R+ - Mild Nudity   

         premiered                                             genres  \
1574   Spring 2006                                               Kids   
1695     Fall 2006                     Adventure, Fantasy, Historical   
1986             -                                 Adventure, Fantasy   
2161     Fall 1995                                       Supernatural   
2418             -                                            Fantasy   
...            ...                                                ...   
18157       

Dari hasil kode diatas kita bisa lihat bahwa:
- Ada 5.296 baris yang mengandung nilai -1 di kolom score.

Lanjut keproses mengubah nilai -1 menjadi Nan.

- Pertama kita buat variabel baru dengan nama `anime_clean` yang akan diisi dengan data_anime dengan kode `anime_clean = data_anime.copy()`.

- Setelah itu kita mengganti nilai -1 dengan fungsi `.replace({-1: np.nan}, inplace=True)`, nilai di dalam `{-1: np.nan}` berarti -1 diganti dengan nilai NaN dan `inplace=True` supaya jika kita berhasil mengganti nilai rating_user mengembalinkan nilai None / tidak ada.

- Setelah itu kita panggil variabel `anime_clean` untuk melihat apakah sudah berubah?

In [9]:
anime_clean = data_anime.copy()
anime_clean["score"].replace({-1: np.nan}, inplace=True)
anime_clean

mal_id                                              title  \
0           1                                       Cowboy Bebop   
1           5                    Cowboy Bebop: Tengoku no Tobira   
2           6                                             Trigun   
3           7                                 Witch Hunter Robin   
4           8                                     Bouken Ou Beet   
...       ...                                                ...   
18157   48952              Twisted Wonderland 1st Anniversary PV   
18158   48953  Natsume Yuujinchou: Hitoyoshi Kuma de no Yasas...   
18159   48954                                     Hairpin Double   
18160   48956                        Wu Shang Shen Di 2nd Season   
18161   48960                                             Hunter   

                                                synopsis background  \
0      In the year 2071, humanity has colonized sever...          -   
1      Another day, another bounty—such is the life o...          -   
2      Vash the Stampede is the man with a $$60,000,0...          -   
3      Witches are individuals with special powers li...          -   
4      It is the dark century and the people are suff...          -   
...                                                  ...        ...   
18157  Disney's Twisted Wonderland mobile game had it...          -   
18158  A collaboration with the Kumamoto prefecture i...          -   
18159                                                  -          -   
18160                                                  -          -   
18161  Hunter is looking for Gossip every night. A sh...          -   

                              aired  airing       duration  episodes     type  \
0       Apr 3, 1998 to Apr 24, 1999       0  24 min per ep        26       TV   
1                       Sep 1, 2001       0    1 hr 55 min         1    Movie   
2       Apr 1, 1998 to Sep 30, 1998       0  24 min per ep        26       TV   
3       Jul 2, 2002 to Dec 24, 2002       0  25 min per ep        26       TV   
4      Sep 30, 2004 to Sep 29, 2005       0  23 min per ep        52       TV   
...                             ...     ...            ...       ...      ...   
18157                  Mar 17, 2021       0          2 min         1  Special   
18158                  Apr 21, 2021       0          2 min         1      ONA   
18159                     2021 to ?       0        Unknown        -1       TV   
18160             May 21, 2021 to ?       1   7 min per ep        64      ONA   
18161                  Sep 14, 2017       0          8 min         1    Movie   

       favorites  ...  score  scored_by                          rating  \
0          59968  ...   8.77     661519  R - 17+ (violence & profanity)   
1           1063  ...   8.39     168515  R - 17+ (violence & profanity)   
2          11882  ...   8.23     288760       PG-13 - Teens 13 or older   
3            512  ...   7.27      37135       PG-13 - Teens 13 or older   
4             10  ...   6.97       5463                   PG - Children   
...          ...  ...    ...        ...                             ...   
18157          0  ...    NaN         -1       PG-13 - Teens 13 or older   
18158          0  ...    NaN         -1                    G - All Ages   
18159          0  ...    NaN         -1                            None   
18160          2  ...    NaN         -1       PG-13 - Teens 13 or older   
18161          0  ...    NaN         -1                R+ - Mild Nudity   

         premiered                                             genres  \
0      Spring 1998    Action, Adventure, Comedy, Drama, Sci-Fi, Space   
1                -              Action, Drama, Mystery, Sci-Fi, Space   
2      Spring 1998  Action, Sci-Fi, Adventure, Comedy, Drama, Shounen   
3      Summer 2002  Action, Mystery, Police, Supernatural, Drama, ...   
4        Fall 2004          Adventure, Fantasy, Shounen, Supernatural   
...            ...       

Hasil di atas menunjukkan bahwa kita berhasil merubah nilai yang tadinya -1 menjadi NaN atau kosong. Sebelum lanjut kita cek lagi apakah nilai -1 nya masih ada atau tidak dengan kode `anime_clean.loc[(anime_clean['score']==-1)]`



In [11]:
anime_clean.loc[(anime_clean['score']==-1)]

Empty DataFrame
Columns: [mal_id, title, synopsis, background, aired, airing, duration, episodes, type, favorites, members, rank, popularity, score, scored_by, rating, premiered, genres, related, status, licensors, producers, studios]
Index: []

[0 rows x 23 columns]

Dari hasil diatas kita bisa yakin bahwa sudah tidak ada nilai -1 di kolom score. Selanjutnya kita akan melanjutkan ke proses selanjutnya.

**2.  Kita akan menghapus data yang kosong.**

Pertama kita cek lagi data yang kosong kita dengan kode `anime_clean.isnull().sum()`.

In [10]:
anime_clean.isnull().sum()

mal_id           0
title            0
synopsis         0
background       0
aired            0
airing           0
duration         0
episodes         0
type             0
favorites        0
members          0
rank             0
popularity       0
score         5296
scored_by        0
rating           0
premiered        0
genres          67
related          0
status           0
licensors        0
producers        0
studios          0
dtype: int64

Dari hasil output diatas kita jadi tau bahwa ada tambahan data kosng yaitu pada kolom score sebanyak 5296.

Setelah berhasil kita hapus data kosong dengan kode `anime_clean = anime_clean.dropna(axis = 0, how ='any')`.

- Dimana `axis = 0` berarti hanya mengapus baris yang mengandung nilai kosong.
- Sedang `how = 'any'` berarti akan menghapus juga jika ada baris atau kolom yang nilainya kosong semua.

In [12]:
anime_clean = anime_clean.dropna(axis = 0, how ='any') 

Setelah kita jalankan kode diatas kita cek lagi apakah datanya yang kosong sudah terhapus? Kita cek dengan kode `anime_clean.isnull().sum()`.

In [13]:
anime_clean.isnull().sum()

mal_id        0
title         0
synopsis      0
background    0
aired         0
airing        0
duration      0
episodes      0
type          0
favorites     0
members       0
rank          0
popularity    0
score         0
scored_by     0
rating        0
premiered     0
genres        0
related       0
status        0
licensors     0
producers     0
studios       0
dtype: int64

Hasilnya sudah tidak ada lagi data kosong di variabel anime_clean.

In [ ]:
import re
def text_cleaning(text):
    text = re.sub(r'&quot;', '', text)
    text = re.sub(r'.hack//', '', text)
    text = re.sub(r'&#039;', '', text)
    text = re.sub(r'A&#039;s', '', text)
    text = re.sub(r'I&#039;', 'I\'', text)
    text = re.sub(r'&amp;', 'and', text)
    
    return text

data_anime['name'] = data_anime['name'].apply(text_cleaning)

# **Modeling**

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer


tfv = TfidfVectorizer(min_df=3,  max_features=None, 
            strip_accents='unicode', analyzer='word',token_pattern=r'\w{1,}',
            ngram_range=(1, 3),
            stop_words = 'english')

# Filling NaNs with empty string
data_anime['genre'] = data_anime['genre'].fillna('')
genres_str = data_anime['genre'].str.split(',').astype(str)
tfv_matrix = tfv.fit_transform(genres_str)

In [ ]:
tfv_matrix.shape

(12294, 1554)

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity
 
# Menghitung cosine similarity pada matrix tf-idf
cosine_sim = cosine_similarity(tfv_matrix) 
cosine_sim

array([[1.        , 0.02312668, 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.02312668, 1.        , 0.02549423, ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.02549423, 1.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 1.        ,
        1.        ],
       [0.        , 0.        , 0.        , ..., 1.        , 1.        ,
        1.        ],
       [0.        , 0.        , 0.        , ..., 1.        , 1.        ,
        1.        ]])

In [ ]:
# Membuat dataframe dari variabel cosine_sim dengan baris dan kolom berupa nama resto
cosine_sim_df = pd.DataFrame(cosine_sim, index=data_anime['name'], columns=data_anime['name'])
print('Shape:', cosine_sim_df.shape)
 
# Melihat similarity matrix pada setiap resto
cosine_sim_df.sample(10, axis=1).sample(10, axis=0)

Shape: (12294, 12294)


name                                          Shin Dokonjou Gaeru: Dokonjou Yumemakura  \
name                                                                                     
Oppai Heart: Kanojo wa Kedamono Hatsujouki!?                                  0.000000   
Huckleberry Finn Monogatari                                                   0.000000   
Sekai Meisaku Douwa Series                                                    0.000000   
New Game!                                                                     1.000000   
Koe wo Kikasete                                                               0.000000   
Maria†Holic                                                                   0.022863   
Hello Kitty no Minna no Tanoshii Natsuyasumi                                  0.000000   
Rayearth                                                                      0.000000   
Omae Umasou da na                                                             0.000000   
Doraemon Movie 05: Nobita no Makai Daibouken                                  0.000000   

name                                          To Heart 2 Special  \
name                                                               
Oppai Heart: Kanojo wa Kedamono Hatsujouki!?            0.000000   
Huckleberry Finn Monogatari                             0.000000   
Sekai Meisaku Douwa Series                              0.000000   
New Game!                                               0.000000   
Koe wo Kikasete                                         0.107843   
Maria†Holic                                             0.058147   
Hello Kitty no Minna no Tanoshii Natsuyasumi            0.000000   
Rayearth                                                0.000000   
Omae Umasou da na                                       0.000000   
Doraemon Movie 05: Nobita no Makai Daibouken            0.000000   

name                                          Mobile Suit Gundam Seed MSV Astray  \
name                                                                               
Oppai Heart: Kanojo wa Kedamono Hatsujouki!?                                 0.0   
Huckleberry Finn Monogatari                                                  0.0   
Sekai Meisaku Douwa Series                                                   0.0   
New Game!                                                                    0.0   
Koe wo Kikasete                                                              0.0   
Maria†Holic                                                                  0.0   
Hello Kitty no Minna no Tanoshii Natsuyasumi                                 0.0   
Rayearth                                                                     0.0   
Omae Umasou da na                                                            0.0   
Doraemon Movie 05: Nobita no Makai Daibouken                                 0.0   

name                                          Beet the Vandel Buster Excellion  \
name                                                                             
Oppai Heart: Kanojo wa Kedamono Hatsujouki!?                          0.000000   
Huckleberry Finn Monogatari                                           0.180446   
Sekai Meisaku Douwa Series                                            0.142147   
New Game!                                                             0.000000   
Koe wo Kikasete                                                       0.000000   
Maria†Holic                                                           0.000000   
Hello Kitty no Minna no Tanoshii Natsuyasumi                          0.083272   
Rayearth                                                              0.251830   
Omae Umasou da na                                                     0.190242   
Doraemon Movie 05: Nobita no Makai Daibouken                          0.181584   

name                                          Moshidora  \
name                                             

In [ ]:
from sklearn.metrics.pairwise import sigmoid_kernel

# Compute the sigmoid kernel
sig = sigmoid_kernel(tfv_matrix, tfv_matrix)

In [ ]:
sig

array([[0.76186428, 0.76160041, 0.76159416, ..., 0.76159416, 0.76159416,
        0.76159416],
       [0.76160041, 0.76186428, 0.76160105, ..., 0.76159416, 0.76159416,
        0.76159416],
       [0.76159416, 0.76160105, 0.76186428, ..., 0.76159416, 0.76159416,
        0.76159416],
       ...,
       [0.76159416, 0.76159416, 0.76159416, ..., 0.76186428, 0.76186428,
        0.76186428],
       [0.76159416, 0.76159416, 0.76159416, ..., 0.76186428, 0.76186428,
        0.76186428],
       [0.76159416, 0.76159416, 0.76159416, ..., 0.76186428, 0.76186428,
        0.76186428]])

In [ ]:
indices = pd.Series(data_anime.index, index=data_anime['name']).drop_duplicates()

In [ ]:
indices

name
Kimi no Na wa.                                            0
Fullmetal Alchemist: Brotherhood                          1
Gintama°                                                  2
Steins;Gate                                               3
Gintama                                                   4
                                                      ...  
Toushindai My Lover: Minami tai Mecha-Minami          12289
Under World                                           12290
Violence Gekiga David no Hoshi                        12291
Violence Gekiga Shin David no Hoshi: Inma Densetsu    12292
Yasuji no Pornorama: Yacchimae!!                      12293
Length: 12294, dtype: int64

In [ ]:
def give_rec(title, sig=sig):
    # Get the index corresponding to original_title
    idx = indices[title]

    # Get the pairwsie similarity scores 
    sig_scores = list(enumerate(sig[idx]))

    # Sort the movies 
    sig_scores = sorted(sig_scores, key=lambda x: x[1], reverse=True)

    # Scores of the 10 most similar movies
    sig_scores = sig_scores[1:6]

    # Movie indices
    anime_indices = [i[0] for i in sig_scores]

    # Top 10 most similar movies
    return pd.DataFrame({'Anime name': data_anime['name'].iloc[anime_indices].values,
                                 'Rating': anime_fulldata['rating_web'].iloc[anime_indices].values,
                                  'episodes': data_anime['episodes'].iloc[anime_indices].values})

In [ ]:
give_rec('Overlord')

Anime name  Rating episodes
0                   Overlord Movie    9.26        1
1                      Log Horizon    9.37       25
2           Log Horizon 2nd Season    9.37       25
3  Sword Art Online II: Debriefing    9.26        1
4            Ys IV: The Dawn of Ys    9.26        1

In [ ]:
lst = [('candy','30','100'), ('apple','10','200'), ('baby','20','300')]
lst.sort(key=lambda x:x[0])
print(lst)

[('apple', '10', '200'), ('baby', '20', '300'), ('candy', '30', '100')]


In [ ]:
# Declare a list of tuples
tuple_list = [("HTML", 15, 'M01'), ("JavaScript", 10, 'M03'), ("Bootstrap", 5, 'M02')]
# Sort the list based on the first item of the tuple
sorted_list1 = sorted(tuple_list, key=lambda x: x[0])
# Print the first sorted list
print("The sorted list based on the first item:\n", sorted_list1)
# Sort the list based on the second item of the tuple
sorted_list2 = sorted(tuple_list, key=lambda x: x[1])
# Print the second sorted list
print("The sorted list based on the second item:\n", sorted_list2)
# Sort the list based on the third item of the tuple
sorted_list3 = sorted(tuple_list, key=lambda x: x[2])
# Print the third sorted list
print("The sorted list based on the third item:\n", sorted_list3)

The sorted list based on the first item:
 [('Bootstrap', 5, 'M02'), ('HTML', 15, 'M01'), ('JavaScript', 10, 'M03')]
The sorted list based on the second item:
 [('Bootstrap', 5, 'M02'), ('JavaScript', 10, 'M03'), ('HTML', 15, 'M01')]
The sorted list based on the third item:
 [('HTML', 15, 'M01'), ('Bootstrap', 5, 'M02'), ('JavaScript', 10, 'M03')]


[enumerate()](https://realpython.com/python-enumerate/#using-pythons-enumerate)